### PR TITLE
[WIP] Role creation and API handling.

### PR DESCRIPTION
**Pull request by RoySegall**
#34

API example:

``` php
if (!$role = OgRole::loadByName('Member', ['gid' => ['value' => 1, 'operator' => '='], 'group_type' => 'node'])) {
  $role = OgRole::create()
    ->setName('Member')
    ->setGid('1')
     ->setGroupType('node');
  $role->save();
}
```

### CODE REVIEW

**Comment by RoySegall** [comment]
Another part in the api:

``` php
if (!$role_permission = OgRolePermission::loadByName('create group_content content', ['module' => 'og', 'permission' => 'create group_content content', 'rid' => $role->id()])) {
  $role_permission = OgRolePermission::create();

  $role_permission
    ->setModule('og')
    ->setRid($role)
    ->setPermission('create group_content content')
    ->save();
}
```

**Comment by damiankloip** [comment]
@RoySegall Do you think OgRole makes sense as content entity? Seems to be more of a configuration entity?

**Comment by RoySegall** [comment]
For me - yes since we're referencing a permission to a role which is a content entity as well because a role referencing to a group.

I don't think we can export the reference to a content entity using configuration management and even if we could it's feel kind of wrong for me. And I do remember @amitaibu, correct me if i'm wrong, don't like to export permission using feature.

Except that - what do you think on the API?

File: src/Entity/OgRole.php [hunk 1]
```diff
@@ -160,22 +155,79 @@ public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
       ->setLabel(t('Role ID'))
       ->setDescription(t('Primary Key: Unique role ID.'));
 
+    $fields['name'] = BaseFieldDefinition::create('string')
+      ->setLabel(t('Name'))
+      ->setDescription(t('Unique role name per group.'));
+
+    // todo: set gid, group_type and group_bundle as null.
+//    <aspilicious_home>     $fields['pass'] = BaseFieldDefinition::create('password')
```
Comment (amitaibu) [inline src/Entity/OgRole.php:163]: aspilicious_home? :) (yeah, I know, WIP :smile: )

File: src/Entity/OgRole.php [hunk 2]
```diff
@@ -160,22 +155,79 @@ public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
...
+    $this
+      ->setGid($entity->id())
+      ->setGroupType($entity->getEntityTypeId())
+      ->setGroupBundle($entity->bundle());
+
+    return $this;
+  }
+
+  /**
+   * Query role IDS.
```
Comment (amitaibu) [inline src/Entity/OgRole.php:204]: => IDs

File: src/Og.php [hunk 3]
```diff
@@ -324,4 +325,69 @@ public static function getSelectionHandler($entity, $bundle, $field_name, array
...
+   *   The name of the main property to query against. i.e. name.
+   * @param $property_value
+   *   The value of the main property.
+   * @param $fields
+   *   Additional fields to the query.
+   *
+   * @return array
+   *   List of entity IDs.
+   */
+  public static function permissionQueryConstructor($realm, $main_property, $property_value, $fields) {
```
Comment (amitaibu) [inline src/Og.php:333]: I don't think this method saves us much.

**Comment by amitaibu** [comment]
> Seems to be more of a configuration entity?

Yes, it should be a _config_ entity. Some time ago I raised my concerns about making it an entity at all instead of a plain DB operations, but since then I saw "everything is an entity" :)

**Comment by amitaibu** [comment]
> Except that - what do you think on the API?

I know it's WIP, but it would help in the review to have a test, in order to see how the API is used.

File: src/Entity/OgRole.php (hunk 1 above)
Comment (RoySegall) [inline src/Entity/OgRole.php:163]: IRC quote for how to enable null property.

**Comment by amitaibu** [comment]
OgRoles are already in.

**Pull request closed without merging**